runtime-modules/project-token/src/lib.rs

```diff
@@ -0,0 +1,703 @@
+        // ensure can freeze amount
+        ensure!(
+            account_info.free_balance >= amount,
+            Error::<T>::InsufficientFreeBalanceForReserving,
+        );
+
+        // == MUTATION SAFE ==
+
+        AccountInfoByTokenAndAccount::<T>::mutate(token_id, &who, |account_data| {
+            account_data.free_balance = account_data.free_balance.saturating_sub(amount);
+            account_data.reserved_balance = account_data.reserved_balance.saturating_add(amount);
+        });
+
+        Self::deposit_event(RawEvent::TokenAmountReservedFrom(token_id, who, amount));
+        Ok(())
+    }
+
+    /// Unreserve `amount` of token for `who`
+    /// Preconditions:
+    /// - `token_id` must id
+    /// - `who` must identify valid account for `token_id`
+    /// - `who` reserved balance must be greater than `amount`
+    ///
+    /// Postconditions:
+    /// - `who` free balance increased by `amount`
+    /// - `who` reserved balance decreased by `amount`
+    /// if `amount` is zero it is equivalent to a no-op
+    fn _unreserve(
+        token_id: T::TokenId,
+        who: T::AccountId,
+        amount: <T as Trait>::Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        // ensure token validity
+        Self::ensure_token_exists(token_id).map(|_| ())?;
+
+        // ensure src account id validity
+        let account_info = Self::ensure_account_data_exists(token_id, &who)?;
+
+        // ensure can freeze amount
+        ensure!(
+            account_info.reserved_balance >= amount,
+            Error::<T>::InsufficientReservedBalance
+        );
+
+        // == MUTATION SAFE ==
+
+        AccountInfoByTokenAndAccount::<T>::mutate(token_id, &who, |account_data| {
+            account_data.free_balance = account_data.free_balance.saturating_add(amount);
+            account_data.reserved_balance = account_data.reserved_balance.saturating_sub(amount);
+        });
+
+        Self::deposit_event(RawEvent::TokenAmountUnreservedFrom(token_id, who, amount));
+
+        Ok(())
+    }
+
+    // Utility ensure checks
+
+    pub(crate) fn ensure_account_data_exists(
+        token_id: T::TokenId,
+        account_id: &T::AccountId,
+    ) -> Result<AccountDataOf<T>, DispatchError> {
+        ensure!(
+            AccountInfoByTokenAndAccount::<T>::contains_key(token_id, account_id),
+            Error::<T>::AccountInformationDoesNotExist,
+        );
+        Ok(Self::account_info_by_token_and_account(
+            token_id, account_id,
+        ))
+    }
+
+    pub(crate) fn ensure_token_exists(
+        token_id: T::TokenId,
+    ) -> Result<TokenDataOf<T>, DispatchError> {
+        ensure!(
+            TokenInfoById::<T>::contains_key(token_id),
+            Error::<T>::TokenDoesNotExist,
+        );
+        Ok(Self::token_info_by_id(token_id))
+    }
+
+    /// Perform token de-issuing: unfallible
+    #[inline]
+    pub(crate) fn do_deissue_token(token_id: T::TokenId) {
```

Comment:
Only allowed if there are no accounts and no balance.

This reminds me that we are not tracking # of active accounts, which we would need to to enforce this, we are only tracking currency units. As my comment above suggest, account deletion only occurs through new extrinsic `dust_account` above.